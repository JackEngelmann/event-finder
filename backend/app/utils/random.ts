export function randomInt(from: number, to: number) {
    return Math.round(Math.random() * (to - from)) + from
}

export function crossProduct<TypeA, TypeB>(a: TypeA[], b: TypeB[]) {
    const results: [TypeA, TypeB][] = []
    for (let elementA of a) {
        for (let elementB of b) {
            const item: [TypeA, TypeB] = [elementA, elementB]
            results.push(item)
        }
    }
    return results
}

export function sample<ListItemType>(list: ListItemType[]) {
    const randomIndex = randomInt(0, list.length - 1)
    return list[randomIndex]
}

export function sampleMultipleWithoutRepition<ListItemType>(list: ListItemType[], amount: number) {
    const result: ListItemType[] = []
    if (amount > list.length) {
        throw Error('you cannot sample more elements than the list has entries')
    }
    while (result.length < amount) {
        const sampled = sample(list)
        if (!result.includes(sampled)) {
            result.push(sampled)
        }
    }
    return result
}