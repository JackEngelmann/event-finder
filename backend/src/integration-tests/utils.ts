import { Database } from '../database/database'

export const genreFragment = `
    id
    name
`

export const eventFragment = `
    club {
        id
        name
    }
    name,
    description,
    date,
    id
    imageUrl,
    priceCategory,
    admissionFee,
    admissionFeeWithDiscount,
    link
    special,
    minimumAge,
    amountOfFloors
    genres {
        ${genreFragment}
    }
`

export const clubFragment = `
    address
    contact
    description
    email
    id
    link
    name
    region
    specials
`

export async function insertTestData(db: Database) {
    await db.run(`
        INSERT INTO club (
            name,
            address,
            region,
            contact,
            email,
            specials,
            description,
            link
        ) VALUES (
            'name',
            'address',
            'region',
            'contact',
            'email',
            'specials',
            'description',
            'link'
        )
    `)
    await db.run(`
        INSERT INTO club (
            name,
            address,
            region,
            contact,
            email,
            specials,
            description,
            link
        ) VALUES (
            'name',
            'address',
            'region',
            'contact',
            'email',
            'specials',
            'description',
            'link'
        )
    `)
    await db.run(`
        INSERT INTO event
        (
            name,
            description,
            date,
            clubId,
            imageUrl,
            priceCategory,
            admissionFee,
            admissionFeeWithDiscount,
            special,
            minimumAge,
            amountOfFloors
        ) 
        VALUES (
            'name',
            'description',
            'date',
            1,
            'imageUrl',
            2,
            10.0,
            20.0,
            'specials',
            18,
            4
        )
    `)
    await db.run(`
        INSERT INTO genre (name) VALUES
            ('genre1'),
            ('genre2'),
            ('genre3')
    `)
    return await db.run(`
        INSERT INTO eventGenre (eventId, genreId) VALUES (1, 1)
    `)
}
