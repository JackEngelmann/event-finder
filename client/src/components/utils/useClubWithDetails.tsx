import gql from 'graphql-tag'
import { Club } from '../../api'
import { useQuery } from '@apollo/react-hooks'

export const CLUB_WITH_DETAILS_QUERY = gql`
    query ClubWithDetails($id: Int!) {
        club(id: $id) {
            address
            contact
            description
            email
            id
            imageUrls
            link
            name
            region
            specials
        }
    }
`
type QueryResult = {
    club: Pick<
        Club,
        | 'address'
        | 'contact'
        | 'description'
        | 'email'
        | 'id'
        | 'imageUrls'
        | 'link'
        | 'name'
        | 'region'
        | 'specials'
    >
}

export function useClubWithDetails(clubId: number) {
    const clubQueryResult = useQuery<QueryResult>(CLUB_WITH_DETAILS_QUERY, {
        variables: { id: clubId }
    })
    const club = clubQueryResult.data && clubQueryResult.data.club
    return [club, clubQueryResult] as const
}
