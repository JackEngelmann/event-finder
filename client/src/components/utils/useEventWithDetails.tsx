import { useQuery } from '@apollo/react-hooks'
import gql from 'graphql-tag'
import { Club, Genre, Event } from '../../api'

export const EVENT_WITH_DETAILS_QUERY = gql`
  query eventQuery($eventId: Int!) {
    event(id: $eventId) {
      admissionFee
      admissionFeeWithDiscount
      amountOfFloors
      club {
        id
        name
      }
      date
      description
      genres {
        id
        name
      }
      id
      imageUrls
      links {
        href
        type
      }
      minimumAge
      name
      priceCategory
      special
    }
  }
`

export type EventWithDetails = Pick<
  Event,
  | 'admissionFee'
  | 'admissionFeeWithDiscount'
  | 'amountOfFloors'
  | 'date'
  | 'description'
  | 'id'
  | 'links'
  | 'imageUrls'
  | 'minimumAge'
  | 'name'
  | 'priceCategory'
  | 'special'
> & {
  club: Pick<Club, 'id' | 'name'>
  genres: Pick<Genre, 'id' | 'name'>[]
}

type QueryData = {
  event: EventWithDetails
}

export function useEventWithDetails(eventId: number) {
  const queryResult = useQuery<QueryData>(EVENT_WITH_DETAILS_QUERY, {
    variables: { eventId: eventId },
  })
  const event = queryResult.data && queryResult.data.event
  return [event, queryResult] as const
}
