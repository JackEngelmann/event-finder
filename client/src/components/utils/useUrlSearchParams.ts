import { useLocation } from "react-router";

export function useUrlSearchParams() {
  return new URLSearchParams(useLocation().search)
}