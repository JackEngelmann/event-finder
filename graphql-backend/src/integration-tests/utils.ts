import { Database } from "sqlite3"

export const genreFragment = `
    id
    name
`

export const eventFragment = `
    club {
        id
        name
    }
    name,
    description,
    date,
    imageUrl,
    priceCategory,
    admissionFee,
    admissionFeeWithDiscount,
    special,
    minimumAge,
    amountOfFloors
    genres {
        ${genreFragment}
    }
`

export const clubFragment = `
    name
    address
    region
    contact
    email
    specials
    description
    link
`

export function insertTestData(db: Database) {
    return new Promise(resolve => {
        db.serialize(() => {
            db.run(`
                    INSERT INTO club (
                        [name],
                        [address],
                        [region],
                        [contact],
                        [email],
                        [specials],
                        [description],
                        [link]
                    ) VALUES (
                        'name',
                        'address',
                        'region',
                        'contact',
                        'email',
                        'specials',
                        'description',
                        'link'
                    )
                `)
            db.run(`
                    INSERT INTO club (
                        [name],
                        [address],
                        [region],
                        [contact],
                        [email],
                        [specials],
                        [description],
                        [link]
                    ) VALUES (
                        'name',
                        'address',
                        'region',
                        'contact',
                        'email',
                        'specials',
                        'description',
                        'link'
                    )
                `)
            db.run(`
                    INSERT INTO event
                    (
                        name,
                        description,
                        date,
                        clubId,
                        imageUrl,
                        priceCategory,
                        admissionFee,
                        admissionFeeWithDiscount,
                        special,
                        minimumAge,
                        amountOfFloors
                    ) 
                    VALUES (
                        'name',
                        'description',
                        'date',
                        1,
                        'imageUrl',
                        2,
                        10.0,
                        20.0,
                        'specials',
                        18,
                        4
                    )
                `)
            db.run(`
                    INSERT INTO genre ([name]) VALUES
                        ('genre1'),
                        ('genre2'),
                        ('genre3')
                `)
            db.run(
                `
                    INSERT INTO eventGenre (eventId, genreId) VALUES (1, 1)
                `,
                () => resolve()
            )
        })
    })
}
