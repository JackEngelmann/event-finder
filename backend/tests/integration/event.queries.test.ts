import {
    eventFragment,
    ApolloTestServer,
    createApolloTestServer,
} from '../utils'
import moment from 'moment'
import { createTestEvent } from '../utils/testBuilders'

const DB_NAME = 'eventquerydb'

let apolloTestServer: ApolloTestServer | undefined

beforeEach(async () => {
    apolloTestServer = await createApolloTestServer({
        isAdmin: false,
        dbName: DB_NAME,
        insertTestData: true,
    })
})

afterEach(async () => {
    if (apolloTestServer) await apolloTestServer.destroy()
    apolloTestServer = undefined
})

describe('event queries: ', () => {
    test('events correctly', async () => {
        const result = await apolloTestServer!.client.query({
            query: `{
                events { ${eventFragment} }
            }`,
        })
        expect(result.data).toBeDefined()
        expect(result).toMatchSnapshot()
    })
    test('existing event correctly', async () => {
        const result = await apolloTestServer!.client.query({
            query: `
                query eventQuery($id: Int!) {
                    event(id: $id) { ${eventFragment} }
                }
            `,
            variables: {
                id: 1,
            },
        })
        expect(result.data).toBeDefined()
        expect(result).toMatchSnapshot()
    })
    test('filter events by club', async () => {
        const apolloTestServer = await createApolloTestServer({
            isAdmin: false,
            dbName: DB_NAME,
            insertTestData: false,
        })
        await createTestEvent()
            .setClubId(1)
            .setName('club-1-1')
            .build(DB_NAME)
        await createTestEvent()
            .setClubId(1)
            .setName('club-1-2')
            .build(DB_NAME)
        await createTestEvent()
            .setClubId(2)
            .setName('club-2-1')
            .build(DB_NAME)
        await createTestEvent()
            .setClubId(2)
            .setName('club-2-2')
            .build(DB_NAME)
        const result = await apolloTestServer.client.query({
            query: `
                query eventsQuery($clubId: Int!) {
                    events(filter: { clubId: $clubId}) {
                        name
                    }
                }
            `,
            variables: {
                clubId: 1,
            },
        })
        expect(result.errors).toBeUndefined()
        expect(result.data!.events.length).toBe(2)
        expect(result.data!.events.some((e: any) => (e.name = 'club-1-1')))
        expect(result.data!.events.some((e: any) => (e.name = 'club-1-2')))
    })
    test('not existing event and return undefined', async () => {
        const result = await apolloTestServer!.client.query({
            query: `
                query eventQuery($id: Int!) {
                    event(id: $id) { ${eventFragment} }
                }
            `,
            variables: {
                id: 2,
            },
        })
        expect(result.data!.event).toBeNull()
    })
})
