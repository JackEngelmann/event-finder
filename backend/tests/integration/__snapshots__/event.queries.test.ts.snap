// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event queries:  events correctly 1`] = `
{
  "data": {
    "events": [
      {
        "admissionFee": 10,
        "admissionFeeWithDiscount": 20,
        "amountOfFloors": 4,
        "club": {
          "id": 1,
          "name": "name",
        },
        "date": "date",
        "description": "description",
        "genres": [
          {
            "id": 1,
            "name": "genre1",
          },
        ],
        "id": 1,
        "imageUrls": [],
        "links": [],
        "minimumAge": 18,
        "name": "name",
        "priceCategory": 2,
        "special": "specials",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`event queries:  existing event correctly 1`] = `
{
  "data": {
    "event": {
      "admissionFee": 10,
      "admissionFeeWithDiscount": 20,
      "amountOfFloors": 4,
      "club": {
        "id": 1,
        "name": "name",
      },
      "date": "date",
      "description": "description",
      "genres": [
        {
          "id": 1,
          "name": "genre1",
        },
      ],
      "id": 1,
      "imageUrls": [],
      "links": [],
      "minimumAge": 18,
      "name": "name",
      "priceCategory": 2,
      "special": "specials",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;
